components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,222 @@
+            addIds(patientsBuilder, patientIds);
+        } catch (final QueryException ex) {
+            logger.warn("Failed to retrieve patient with external id [{}]: {}", eid, ex.getMessage());
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of internal patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param ids a list of patient ids, as strings
+     */
+    private void addIds(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> ids)
+    {
+        for (final Object id : ids) {
+            if (StringUtils.isNotBlank((String) id)) {
+                addPatientFromId(patientsBuilder, id);
+            }
+        }
+    }
+
+    /**
+     * Given the patient's internal ID, retrieves the patient entity, if it exists and if the user has view rights, and
+     * adds it to the set of patient entities.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param id an internal patient ID
+     */
+    private void addPatientFromId(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object id)
+    {
+        try {
+            // Try to get the patient entity.
+            final PrimaryEntity patient = this.repository.get((String) id);
+            // Get the current user.
+            final User currentUser = this.users.getCurrentUser();
+            // If user has view rights and patient with the provided ID exists, add patient to patient set.
+            if (patient != null && this.access.hasAccess(Right.VIEW, currentUser == null ? null
+                : currentUser.getProfileDocument(), patient.getDocument())) {
+                patientsBuilder.add(patient);
+            }
+        } catch (final IllegalArgumentException ex) {
+            logger.warn("Failed to retrieve patient with ID [{}]: {}", id, ex.getMessage());
+        }
+    }
+
+    /**
+     * A custom object mapper to facilitate serializing a list of {@link Patient} objects.
+     *
+     * @return an object mapper that can serialize {@link Patient} objects
+     */
+    private static ObjectMapper getCustomObjectMapper()
+    {
+        final ObjectMapper objectMapper = new ObjectMapper();
+        SimpleModule m = new SimpleModule(SERIALIZER_LABEL, new Version(1, 0, 0, "", SERIALIZER_LABEL, "json"));
+        m.addSerializer(PrimaryEntity.class, new PrimaryEntitySerializer());
+        objectMapper.registerModule(m);
+        return objectMapper;
+    }
+
+    /**
+     * A custom serializer for primary entities.
+     */
+    private static final class PrimaryEntitySerializer extends JsonSerializer<PrimaryEntity>
+    {
+        @Override
+        public void serialize(final PrimaryEntity primaryEntity, final JsonGenerator jgen,
+            final SerializerProvider provider) throws IOException
+        {
+            jgen.writeRawValue(primaryEntity.toJSON().toString());
```

Comment:
I'd like to see the same links that are included in [the normal patient JSON](http://localhost:8080/rest/patients/P0000001). This does the trick:

    json.put("links", this.autolinker.get().forSecondaryResource(PatientResource.class, uriInfo)
        .withExtraParameters("patient-id", patient.getId()).build());